src/gridcoin/staking/status.cpp

```diff
@@ -312,3 +316,110 @@ void EfficiencyReport::UpdateMetrics(uint64_t weight_sum, int64_t balance_weight
     // actual_cumulative_weight / ideal_cumulative_weight provides a measure of
     // the overall mining efficiency compared to ideal.
 }
+
+// -----------------------------------------------------------------------------
+// Class: BlockChainStatus
+// -----------------------------------------------------------------------------
+
+BlockChainStatus::BlockChainStatus()
+    : m_last_update_ms(0)
+    , m_cached_num_blocks(-1)
+    , m_cached_num_blocks_of_peers(-1)
+    , m_cached_best_block_time(-1)
+    , m_cached_difficulty(-1)
+    , m_cached_net_weight(-1)
+    , m_cached_etts_days(0)
+{
+// Nothing to do.
+}
+
+void BlockChainStatus::UpdateBlockChainStatus(int height, int64_t best_time)
+{
+    int64_t now = GetTimeMillis();
+
+    // Here we use the model update delay interval for the caching update interval, because it makes sense to align them.
+    if (now - m_last_update_ms < MODEL_UPDATE_DELAY) return;
+
+    m_cached_num_blocks = height;
+    m_cached_best_block_time = best_time;
+
+    {
+        // This should normally be called when cs_main is already held, in which case this is going to succeed.
+        // If called from other thread where cs_main isn't held by the caller, and cs_main is held by another thread,
+        // then this will not block but simply return immediately with nothing done.
+        //
+        // The normal update from this from the core side is a call in the GridcoinServices() function in main.cpp.
+        TRY_LOCK(cs_main, lockMain);
```

Comment:
You may want to use it out of a thread where cs_main is already held. AssertLockHeld(cs_main) is not appropriate for that use case. This works either in a thread where cs_main is already held (an additional harmless superfluous lock is taken), or TRIED in a thread where cs_main is not held already. That was the thinking.

Reply:
Ah, that's a good point. I suppose it could be called from RPC (as described in the other PR) as a last-minute attempt to update the cache if we use this as the source for that data.